x/mongo/driver/topology/connection.go

```diff
@@ -295,16 +295,20 @@ func (c *connection) close() error {
 	if atomic.LoadInt32(&c.connected) != connected {
 		return nil
 	}
-	if c.pool == nil {
-		var err error
 
-		if c.nc != nil {
-			err = c.nc.Close()
+	if c.pool != nil {
+		err := c.pool.removeConnection(c)
```

Comment:
So, the previous call to `c.pool.closeConnection` did a lot of other things too, such as closing the underlying connection (`c.nc.Close()`)... do we need to be doing at least some of that as well?

Reply:
The old call to c.pool.closeConnection would try to remove the connectionID first, and if that worked would close the underlying connection, which we're still doing.

Reply:
Ah, I see it now. I was having some trouble reading the diff, but this conditional calls `c.pool.removeConnection` and the code below closes `c.nc`. The `removeConnection` call really shouldn't ever error because the only error is if the pool doesn't match this connection's pool, but would it make more sense logically to close `nc` first? That way, this function guarantees that the underlying resources are always closed, even if an error occurs.

Reply:
@divjotarora done